### PR TITLE
Add 90 degree rotation methods.

### PR DESCRIPTION
**Pull request by Tetragramm**
Adds a method for fast 90 degree rotations.  It is a common task, but there is no method for it.  It can be done with just transpose and flips, but this saves people from having to try each time to get it right.

### CODE REVIEW

**Comment by alalek** [comment]
What coordinate system is assumed here?

BTW, In general, CCW rotation is [used](https://en.wikipedia.org/wiki/Angle_of_rotation) instead of CW.

/cc @vpisarev

**Comment by Tetragramm** [comment]
Perhaps I should have made it more clear.  It's fast 90 degree rotations of an image.  Like flip and mirror, this is one of the very basic operations.  It just annoyed me one too many times that there wasn't a method for it, so I wrote one.  ROTATE_90, the top side of the image becomes the right side, ROTATE_180, the top becomes the bottom, ROTATE_270, the top becomes the left side.

The comments specify that it's clockwise, but it's simple enough to re-name the enums if that's preferred.  Or it could be Rotate 90 Left and 90 Right like some programs use.

**Comment by StevenPuttemans** [comment]
> What coordinate system is assumed here?

I would plea to stick to the OpenCV coordinate system, which is top left corner as origin, x axis from left to right and y axis from top to bottom. Also lets mention it specificly in the function, since some other functions tend to do it otherwise and thus complicate things.

Also I would then go clockwise in that coordinate system.

**Comment by Tetragramm** [comment]
[THIS](http://www.computerhope.com/jargon/r/rotate.gif) is the convention currently used.  We could reverse it to be all CCW and switch the 90 and 270 in that picture, or change it to CW_90 and CCW_90 which is unambiguous.

I'm starting to think being unambiguous is the way to go.  Either way, no matter how you set the coordinate system, clockwise always means the same thing.

Does everyone think CW_90 and CCW_90 are the way to go?

File: modules/core/include/opencv2/core.hpp [hunk 1]
```diff
@@ -1022,6 +1022,24 @@ around both axes.
...
+*   Rotate by 90 degrees clockwise (rotateCode = ROTATE_90).
+*   Rotate by 180 degrees clockwise (rotateCode = ROTATE_180).
+*   Rotate by 270 degrees clockwise (rotateCode = ROTATE_270).
+@param src input array.
+@param dst output array of the same type as src.  The size is the same with ROTATE_180,
+and the rows and cols are switched for ROTATE_90 and ROTATE_270.
+@param rotateCode an enum to specify how to rotate the array; see the enum RotateFlags
+@sa transpose , repeat , completeSymm, flip, RotateFlags
+*/
+CV_EXPORTS_W void rotate(InputArray src, OutputArray dst, int rotateCode);
```
Comment (vpisarev) [inline modules/core/include/opencv2/core.hpp]: can we replace rotateCode with angle and simple check that it's divisible by 90? (and also allow negative values?)

File: modules/core/src/copy.cpp [hunk 2]
```diff
@@ -826,6 +826,50 @@ void flip( InputArray _src, OutputArray _dst, int flip_mode )
...
 }
 
+#ifdef HAVE_OPENCL
+
+static bool ocl_rotate(InputArray _src, OutputArray _dst, int rotateMode)
+{
+    switch (rotateMode)
+    {
+    case ROTATE_90:
+        flip(_src.getUMat().t(), _dst, 1);
```
Comment (vpisarev) [inline modules/core/src/copy.cpp]: this implementation is inefficient, it allocates a temporary array for transposed image. More efficiently would be to write it as:

<pre>
transpose(_src, _dst);
flip(_dst, _dst, 1);
</pre>

File: modules/core/src/copy.cpp [hunk 3]
```diff
@@ -826,6 +826,50 @@ void flip( InputArray _src, OutputArray _dst, int flip_mode )
...
+        break;
+    case ROTATE_180:
+        flip(_src, _dst, -1);
+        break;
+    case ROTATE_270:
+        flip(_src.getUMat().t(), _dst, 0);
```
Comment (vpisarev) [inline modules/core/src/copy.cpp]: same as above

File: modules/core/src/copy.cpp [hunk 4]
```diff
@@ -826,6 +826,50 @@ void flip( InputArray _src, OutputArray _dst, int flip_mode )
...
+void rotate(InputArray _src, OutputArray _dst, int rotateMode)
+{
+    CV_Assert(_src.dims() <= 2);
+
+    CV_OCL_RUN(_dst.isUMat(), ocl_rotate(_src, _dst, rotateMode))
+
+    switch (rotateMode)
+    {
+    case ROTATE_90:
+        flip(_src.getMat().t(), _dst, 1);
```
Comment (vpisarev) [inline modules/core/src/copy.cpp]: same issue with efficiency

File: modules/core/src/copy.cpp [hunk 5]
```diff
@@ -826,6 +826,50 @@ void flip( InputArray _src, OutputArray _dst, int flip_mode )
...
+        break;
+    case ROTATE_180:
+        flip(_src, _dst, -1);
+        break;
+    case ROTATE_270:
+        flip(_src.getMat().t(), _dst, 0);
```
Comment (vpisarev) [inline modules/core/src/copy.cpp]: same issue with efficiency

**Comment by vpisarev** [comment]
@Tetragramm, @alalek, I'm fine with the direction of rotation and interpretation of angles, for me it looks natural. But I would use angle value directly, as I suggested above

**Comment by Tetragramm** [comment]
Excellent point on the temporary matrix.  I think I'm spoiled by Eigen where a transpose is just a flag.

Should we really do a number angle?  If I saw that I would expect to be able to put in any number and get a rotated image.  I would think an enum with all the possible options would be best.  I will change it to 90 Clockwise and 90 Counter Clockwise though.

File: modules/core/src/copy.cpp [hunk 6]
```diff
@@ -826,6 +826,54 @@ void flip( InputArray _src, OutputArray _dst, int flip_mode )
...
+    {
+    case ROTATE_90_CLOCKWISE:
+        _dst.getMat() = _src.getMat().t();
+        flip(_dst, _dst, 1);
+        break;
+    case ROTATE_180:
+        flip(_src, _dst, -1);
+        break;
+    case ROTATE_90_COUNTERCLOCKWISE:
+        _dst.getMat() = _src.getMat().t();
```
Comment (vpisarev) [inline modules/core/src/copy.cpp]: why not use transpose() function?

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**